cases/src/org/commcare/cases/util/CasePurgeFilter.java

```diff
@@ -222,7 +227,96 @@ private void propogateMarkToDAG(DAG<String, int[], String> dag, boolean walkFrom
+            for (Edge<String, String> edge : edgeListForOrigin) {
+                String targetIndex = edge.i;
+                if (internalCaseDAG.getNode(targetIndex) == null) {
+                    invalidEdgesRemoved = true;
+                    edgesToRemove.add(new String[]{originIndex, targetIndex});
+                    childOfNonexistentParent.add(originIndex);
+                }
+            }
+        }
+
+        // Any case node with a nonexistent parent should be removed from the graph, as should
+        // all of its descendants, and any edges to or from it
+        for (String index : childOfNonexistentParent) {
+            edgesToRemove = removeNodeAndPropagate(index, edgesToRemove);
+        }
+
+        // Do actual edge removal after we finish all iterations through the edge list, to prevent
+        // ConcurrentModificationException
+        for (String[] edge : edgesToRemove) {
+            internalCaseDAG.removeEdge(edge[0], edge[1]);
+        }
+
+        return invalidEdgesRemoved;
+    }
+
+    /**
+     * Remove from the graph the node at this index, and remove all nodes that are made
+     * invalid by the non-existence of that node (i.e. all of its child and extension cases)
+     *
+     * @return A list of edges that need to be removed as a result of this method call. Actual edge
+     * removal should then be done all at once by the caller, to avoid
+     * ConcurrentModificationException
+     */
+    private Set<String[]> removeNodeAndPropagate(String indexOfRemovedNode,
+                                        Set<String[]> accumulatedEdgesToRemove) {
+        // Wording is confusing here -- because all edges in this graph are from a child case to
+        // a parent case, calling getParents() for a node returns all of its child/extension cases
+        Vector<Edge<String, String>> childCases = internalCaseDAG.getParents(indexOfRemovedNode);
+        for (Edge<String, String> child : childCases) {
+            // Want to remove the edge from child case to parent case
+            accumulatedEdgesToRemove.add(new String[]{child.i, indexOfRemovedNode});
+
+            // Recurse on child case
+            accumulatedEdgesToRemove = removeNodeAndPropagate(child.i, accumulatedEdgesToRemove);
+        }
+
+        // Also want to remove any outgoing edges to other parents (but do NOT delete those
+        // parents because they are still valid)
+        Vector<Edge<String, String>> parentCases = internalCaseDAG.getChildren(indexOfRemovedNode);
+        for (Edge<String, String> parent : parentCases) {
+            accumulatedEdgesToRemove.add(new String[]{indexOfRemovedNode, parent.i});
+        }
+
+        // Once all edges to/from this node have been removed, delete the node itself from the
+        // DAG, and add it to the list of cases to be purged
+        int storageIdOfRemovedNode = internalCaseDAG.getNode(indexOfRemovedNode)[1];
+        internalCaseDAG.removeNode(indexOfRemovedNode);
```

Comment:
If you make `removeNode` have the same type signature as `Hashtable.remove` then you don't need to call `getNode` in the line above but rather:
```
int storageIdOfRemovedNode = internalCaseDAG.removeNode(indexOfRemovedNode)[1];
```